Assignment 1 Solution

shuveksha-tuladhar opened the pull request:
## Description

This PR includes the solution to assignment 1 which includes the concepts of data types, operators, control flow statements, and functions. Each task addresses a specific task related to string manipulation, arithmetic operations, iteration, and error handling. 

## Tasks:

1. Hello Function (hello)  
    - Returns` "Hello!" without any parameters.
2. Greet Function (greet)
    - Accepts a name and returns using a formatted string.
3. Calculator Function (calc)
    - Takes two numbers and an operation (add, subtract, multiply, divide, modulo, int_divide, power).
    - Handles division by zero and invalid operations with appropriate error messages.
4. Data Type Conversion (data_type_conversion)
    - Converts a value to int, float, or str and handles conversion errors gracefully.
5. Grading System (grade)
    - Uses *args to calculate an average grade and returns an appropriate letter grade (A–F) and handles invalid inputs.
6. String Repetition (repeat)
    - Uses a for loop and range() to repeat a string a given number of times.
7. Student Scores (student_scores)
    - Uses **kwargs to return the highest score (best) or the average (mean).
8. Title Case Formatter (titleize)
    - Capitalizes a string like a book title while keeping certain small words lowercase (a, an, the, of, and, is, in, on).
9. Hangman (hangman)
    - Replaces unguessed letters with _ in a secret word based on a given guess string.
10. Pig Latin (pig_latin)
    - Converts words into Pig Latin following the correct transformation rules.

Ifedolapo-9 commented:
Hello! Great work.

Here are my comments
  
Generally, calling functions can be done at once at the end of the script: 

1.  Fix: return "Hello!"
Returning a value doesn’t require parentheses unless explicitly needed.

4. Here's a simpler version: 

def calc(num1, num2, operation):
    operations = {
        "add": num1 + num2,
        "subtract": num1 - num2,
        "multiply": num1 * num2,
        "divide": num1 / num2 if num2 != 0 else "Error: Division by zero",
        "modulo": num1 % num2 if num2 != 0 else "Error: Modulo by zero",
        "int_divide": num1 // num2 if num2 != 0 else "Error: Integer division by zero",
        "power": num1 ** num2
    }
    return operations.get(operation, "Error: Invalid operation")

Plus:
 Default operation is "multiply," which is not clearly indicated inside the function.
Error message "You can't multiply those values!" appears even if multiplication is not chosen.



5. 
  Calling functions can be done at once at the end of the script: 
case _:
    return f"Invalid data type: {data_type}"


8. return best_student, highest_score (To return both name and score)

9. The function keeps small words lowercase but doesn’t properly capitalize them at the beginning.

Overall you did a good work.
Thank you for your time.

shuveksha-tuladhar commented:
@Ifedolapo-9 Thank you for your feedback. I agree that calling functions can be done all at once, but I was running the test file as I proceeded with each task individually.

- I have fixed Task 1 and implemented your suggestions for Task 2.
- Regarding Task 5, I wasn’t entirely sure about your comment, as it seems unrelated to that task.
- The best_student and highest_score are for Task 7. When we return two values, they are returned as tuples, and that approach won’t pass the test case.
![image](https://github.com/user-attachments/assets/0e9f4571-09b2-463b-a7ba-dd0553066039)
- For Task 9, I’ve added the lower() method as you suggested.

Thank you again for your help!

Ifedolapo-9 commented:
Hello,
Thanks for the feedback.
I will give more perspective to Task 5 later.
You are welcome.
Thank you.

On Mon, 24 Mar 2025, 01:47 Shuveksha Tuladhar, ***@***.***>
wrote:

> @Ifedolapo-9 <https://github.com/Ifedolapo-9> Thank you for your
> feedback. I agree that calling functions can be done all at once, but I was
> running the test file as I proceeded with each task individually.
>
>    - I have fixed Task 1 and implemented your suggestions for Task 2.
>    - Regarding Task 5, I wasn’t entirely sure about your comment, as it
>    seems unrelated to that task.
>    - The best_student and highest_score are for Task 7. When we return
>    two values, they are returned as tuples, and that approach won’t pass the
>    test case.
>    image.png (view on web)
>    <https://github.com/user-attachments/assets/0e9f4571-09b2-463b-a7ba-dd0553066039>
>    - For Task 9, I’ve added the lower() method as you suggested.
>
> Thank you again for your help!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/shuveksha-tuladhar/ctd-python-homework/pull/1#issuecomment-2746608292>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZEUXC4LSB3YNSU5A4ZYW632V5I2PAVCNFSM6AAAAABY7LQ3FSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBWGYYDQMRZGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: shuveksha-tuladhar]*shuveksha-tuladhar* left a comment
> (shuveksha-tuladhar/ctd-python-homework#1)
> <https://github.com/shuveksha-tuladhar/ctd-python-homework/pull/1#issuecomment-2746608292>
>
> @Ifedolapo-9 <https://github.com/Ifedolapo-9> Thank you for your
> feedback. I agree that calling functions can be done all at once, but I was
> running the test file as I proceeded with each task individually.
>
>    - I have fixed Task 1 and implemented your suggestions for Task 2.
>    - Regarding Task 5, I wasn’t entirely sure about your comment, as it
>    seems unrelated to that task.
>    - The best_student and highest_score are for Task 7. When we return
>    two values, they are returned as tuples, and that approach won’t pass the
>    test case.
>    image.png (view on web)
>    <https://github.com/user-attachments/assets/0e9f4571-09b2-463b-a7ba-dd0553066039>
>    - For Task 9, I’ve added the lower() method as you suggested.
>
> Thank you again for your help!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/shuveksha-tuladhar/ctd-python-homework/pull/1#issuecomment-2746608292>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZEUXC4LSB3YNSU5A4ZYW632V5I2PAVCNFSM6AAAAABY7LQ3FSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBWGYYDQMRZGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
